indexer/blockapi.go

```diff
@@ -0,0 +1,169 @@
+package indexer
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sync"
+
+	blocksapi "github.com/aurora-is-near/borealis-prototypes-go/generated/blocksapi"
+	"github.com/aurora-is-near/relayer2-base/broker"
+	"github.com/aurora-is-near/relayer2-base/db"
+	"github.com/aurora-is-near/relayer2-base/log"
+	"github.com/aurora-is-near/relayer2-base/types"
+	"github.com/aurora-is-near/relayer2-base/types/common"
+	"github.com/aurora-is-near/relayer2-base/types/indexer"
+	"github.com/aurora-is-near/relayer2-base/utils"
+	jsoniter "github.com/json-iterator/go"
+
+	vtgrpc "github.com/planetscale/vtprotobuf/codec/grpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/encoding"
+	_ "google.golang.org/grpc/encoding/proto"
+	"google.golang.org/grpc/metadata"
+)
+
+type IndexerBlocksAPI struct {
+	token  string
+	dbh    db.Handler
+	l      *log.Logger
+	b      broker.Broker
+	grpc   *grpc.ClientConn
+	height uint64
+	mu     sync.Mutex
+}
+
+func init() {
+	// Register protobuf plugin for generate optimized marshall & unmarshal code
+	encoding.RegisterCodec(vtgrpc.Codec{})
+}
+
+func NewIndexerBlocksApi(config *Config, dbh db.Handler, b broker.Broker) (*IndexerBlocksAPI, error) {
+	logger := log.Log()
+
+	client, err := grpc.NewClient(
+		config.BlockApiUrl,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithInitialConnWindowSize(64*1024*1024),                           // 64 MB
+		grpc.WithInitialWindowSize(64*1024*1024),                               // 64 MB
+		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(1*1024*1024*1024)), // 1GB
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &IndexerBlocksAPI{
+		token: config.BlockApiToken,
+		dbh:   dbh,
+		l:     logger,
+		b:     b,
+		grpc:  client,
+		height: config.FromBlock,
+	}, nil
+}
+
+func (i *IndexerBlocksAPI) Start(ctx context.Context) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+
+	// Attaching authorization token
+	md := metadata.New(make(map[string]string))
+	md.Set("authorization", "Bearer "+i.token)
+
+	callCtx := metadata.NewOutgoingContext(ctx, md)
+
+	go i.run(callCtx)
```

Comment:
Something like:
```golang
wg.Add(1)
go func() {
defer wg.Done()
for {
	if indexerCtx.Err() != nil {
		return
	}
	if err := run(indexerCtx); err == nil {
		// Maybe we finished completely (if !force-reindex && toblock != 0)
		// in that case just return
		// otherwise handle switch from reindex to normal endless indexing
	}
	// TODO: Log error

	// That function doesn't exist but I guess you understand what I mean
	if !sleepCtx(indexerCtx, time.Second/5) {
		return
	}
}
}()
```